test/unit/authorizer_test.rb

```diff
@@ -10,169 +10,203 @@ def setup
-    permission = Permission.find_by_name('view_domains')
-    FactoryGirl.create(:filter, :role => @role, :permissions => [permission])
-    domain     = FactoryGirl.create(:domain)
-    auth       = Authorizer.new(FactoryGirl.create(:user))
-
-    result = auth.find_collection(Domain, :permission => :view_domains)
-    assert_not_includes result, domain
-    assert_kind_of ActiveRecord::Relation, result
-    refute auth.can?(:view_domains, domain)
-  end
-
-  test "#can? caches results per permission and class" do
-    permission1 = Permission.find_by_name('view_domains')
-    FactoryGirl.create(:filter, :on_name_starting_with_a,
-                                 :role => @role, :permissions => [permission1])
-    domain1     = FactoryGirl.create(:domain, :name => 'a-domain.to-be-found.com')
-    domain2     = FactoryGirl.create(:domain, :name => 'x-domain.not-to-be-found.com')
-    permission2 = Permission.find_by_name('view_architectures')
-    architecture = FactoryGirl.create(:architecture)
-    FactoryGirl.create(:filter, :role => @role, :permissions => [permission2])
-
-    auth = Authorizer.new(@user)
-
-    auth.stubs(:find_collection).returns([domain1]).times(3)
-    assert auth.can?(:view_domain, domain1)
-    refute auth.can?('view_domain', domain2)
-    assert auth.can?(:edit_domain, domain1)
-    refute auth.can?('edit_domain', domain2)
-    refute auth.can?(:view_architectures, architecture) # since it's stubbed and returns domain1 only
-    refute auth.can?('view_architectures', architecture)
+  describe '#can?' do
+    [true, false].each do |cache|
+      context "with cache = #{cache}" do
+        # limited, unlimited, permission with resource, without resource...
+        test "with unlimited filter" do
+          FactoryGirl.create(:filter, :role => @role, :permissions => [@permission])
+          auth = Authorizer.new(@user)
+
+          assert auth.can?(@permission.name.to_sym, nil, cache)
+          refute auth.can?(:view_domains, nil, cache)
+        end
+
+        test "with limited filter (name ~ *)" do
+          FactoryGirl.create(:filter, :on_name_all, :role => @role, :permissions => [@permission])
+          auth = Authorizer.new(@user)
+
+          assert auth.can?(@permission.name.to_sym, nil, cache)
+          refute auth.can?(:view_domains, nil, cache)
+        end
+
+        test "permission without resource" do
+          FactoryGirl.create(:filter, :on_name_all, :role => @role, :permissions => [@permission])
+          auth = Authorizer.new(@user)
+
+          assert auth.can?(@permission.name.to_sym, nil, cache)
+          refute auth.can?(:view_domains, nil, cache)
+        end
+
+        test "limited by particular user" do
+          FactoryGirl.create(:filter, :on_name_all, :role => @role, :permissions => [@permission])
+          auth = Authorizer.new(FactoryGirl.create(:user))
+
+          refute auth.can?(@permission.name.to_sym, nil, cache)
+        end
+
+        test "for unlimited filter" do
```

Comment:
These tests that passed a second subject argument through have lost that info, they previously showed `#can?(:view_domains, @host)` which is different to the "with unlimited filter" test above.  It's not very clear from the test name now, even with the describe/context blocks.